### PR TITLE
Sqlservertime

### PR DESCRIPTION
**Pull request by smootoo**

### CODE REVIEW

**Comment by smootoo** [comment]
As requested. A new PR based on #1825 
Don't merge this one yet. I'll fix the Oracle tests first.

BTW, do we still care about the CircleCI build. It looks like it's failing on master as well. I could fix that too?

**Comment by hvesalai** [comment]
I must say I don't know when the circleci tests are run. I have not been following them.

**Comment by smootoo** [comment]
Just a quick update on where I am with this. I got the Oracle profile working with the existing tests, but as I was reviewing the code in the profile, I was confused as to how it could be working. Turns out it didn’t work :-) There is a gap in the test coverage. In particular around updating ResultSets and what’s happening at DST change overs. I’m increasing the test coverage and it’s highlighted some issues with the other profiles as well. I’ll work through those and update this PR. I should have something by this weekend.

**Comment by smootoo** [comment]
Another quick update on where I am. The new tests I added highlighted a few issues, so I've been working through them. It's taken a while, but I'm getting close now. Only the DB2 and postgres tests to fix. Hopefully have them done and the code tidied up in a few days time.

**Comment by hvesalai** [comment]
Thank you for the update. Good progress!

**Comment by smootoo** [comment]
Almost there now. This has turned into a bit of an epic! Fixing the postgres issues were particularly tricksy.
All the builds are green now though. I'll re-review the code in its entirety and beef up some of the comments as I've been heads-down fixing one issue after another.

File: slick/src/main/scala/slick/jdbc/DB2Profile.scala [hunk 1]
```diff
@@ -180,6 +185,70 @@ trait DB2Profile extends JdbcProfile {
       override def sqlTypeName(sym: Option[FieldSymbol]) = "CHAR(1)"
       override def valueToSQLLiteral(value: Boolean) = if(value) "1" else "0"
     }
+
+    class InstantJdbcType extends super.InstantJdbcType {
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(6) WITH TIME ZONE"
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/DB2Profile.scala]: Why would Instant have a time zone? Instant is just time steps since `1970-01-01T00:00:00Z`, i.e. it is always Zulu time

File: slick/src/main/scala/slick/jdbc/DerbyProfile.scala [hunk 2]
```diff
@@ -240,6 +243,27 @@ trait DerbyProfile extends JdbcProfile {
       override def sqlType = java.sql.Types.BINARY
       override def sqlTypeName(sym: Option[FieldSymbol]) = "CHAR(16) FOR BIT DATA"
     }
+
+    class InstantJdbcType extends super.InstantJdbcType {
+      // Derby has no timestamp with timezone type and so using strings as timestamps are
+      // susceptible to DST gaps twice a year
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/DerbyProfile.scala]: this comment should not apply to Instant, which should be stored in Z-time, i.e. without DST

File: slick/src/main/scala/slick/jdbc/DerbyProfile.scala [hunk 3]
```diff
@@ -240,6 +243,27 @@ trait DerbyProfile extends JdbcProfile {
...
+        r.getString(idx) match {
+          case null => null
+          case instantString => Instant.parse(instantString)
+        }
+      }
+      override def updateValue(v: Instant, r: ResultSet, idx: Int): Unit = {
+        r.updateString(idx, v.toString)
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        s"'${value.toString}'"
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/DerbyProfile.scala]: Here the `toString` of Instant is defined as returning a string in the UTC-timezone.

File: slick/src/main/scala/slick/jdbc/H2Profile.scala [hunk 4]
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
       override def valueToSQLLiteral(value: UUID) = "'" + value + "'"
       override def hasLiteralForm = true
     }
+    override val instantType : InstantJdbcType = new InstantJdbcType {
+      val formatter = new DateTimeFormatterBuilder()
+                      .append(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
+                      .optionalStart()
+                      .appendFraction(ChronoField.NANO_OF_SECOND, 0, 9, true)
+                      .optionalEnd()
+                      .appendOffset("+HH", "")
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/H2Profile.scala]: Ditto here. Instant should not be stored only in the UTC zoneoffset, which may also be implied.

File: slick/src/main/scala/slick/jdbc/H2Profile.scala [hunk 5]
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
...
+                      .toFormatter()
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(9) WITH TIME ZONE"
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/H2Profile.scala]: As instants should always be in the UTC zone offset, storing the timezone is not necessary

File: slick/src/main/scala/slick/jdbc/H2Profile.scala [hunk 6]
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
...
+
+      override def setValue(v: Instant, p: PreparedStatement, idx: Int) : Unit = {
+        p.setString(idx, if (v == null) null else v.toString)
+      }
+      override def getValue(r: ResultSet, idx: Int) : Instant = {
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/H2Profile.scala]: this is different from the DerbyProfile, which uses `Instant.parse` Why are they different?

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala [hunk 7]
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
...
+          case null => null
+          case hsqldbString =>
+            @inline val normalizedIsoString: String = offsetConvertHsqldbToISO(hsqldbString)
+            OffsetDateTime.parse(normalizedIsoString,datetimeFormatter)
+        }
+      }
+    }
+    class InstantJdbcType extends super.InstantJdbcType with HsqldbTimeJdbcTypeWithOffset {
+      override def sqlType = java.sql.Types.TIMESTAMP_WITH_TIMEZONE
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(9) WITH TIME ZONE"
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: same comment as with the other Instance types... no zone needed if we expect UTC as implied.

File: slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala [hunk 8]
```diff
@@ -164,12 +172,196 @@ trait JdbcTypesComponent extends RelationalTypesComponent { self: JdbcProfile =>
...
+    }
+
+    class InstantJdbcType extends DriverJdbcType[Instant] {
+      // TODO: Use native Instant JDBC methods as soon they are implemented.
+      // https://java.net/jira/browse/JPA_SPEC-63
+      override def sqlType : Int = {
+        java.sql.Types.TIMESTAMP
+      }
+      override def setValue(v: Instant, p: PreparedStatement, idx: Int) : Unit = {
+        p.setTimestamp(idx, Timestamp.from(v))
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala]: This looks nice. Is there a reason why the other Profiles don't use this method?

File: slick/src/main/scala/slick/jdbc/DB2Profile.scala [hunk 9]
```diff
@@ -176,6 +181,30 @@ trait DB2Profile extends JdbcProfile {
...
+        r.getString(idx) match {
+          case null => null
+          case instantStr => Instant.parse(instantStr)
+        }
+      }
+      override def updateValue(v: Instant, r: ResultSet, idx: Int) : Unit = {
+        r.updateString(idx, v.toString)
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        s"'$value'"
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/DB2Profile.scala]: please add `.toString`

File: slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala [hunk 10]
```diff
@@ -164,12 +172,196 @@ trait JdbcTypesComponent extends RelationalTypesComponent { self: JdbcProfile =>
...
+      }
+      override def getValue(r: ResultSet, idx: Int) : Instant = {
+        r.getTimestamp(idx) match {
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala]: Also looks very clean.

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 11]
```diff
@@ -255,12 +258,191 @@ trait PostgresProfile extends JdbcProfile {
...
+          .appendLiteral("+00")
+          .optionalEnd()
+          .toFormatter()
+      }
+
+      implicit val min : Instant = Instant.MIN
+      implicit val max : Instant = Instant.MAX
+      implicit val serializeFiniteTime : (Instant => String) =  _.toString
+      implicit val parseFiniteTime : (String => Instant) = {
+        LocalDateTime.parse(_, formatter).toInstant(ZoneOffset.UTC)
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: Is this really safe? Here you convert the string from the database to the Local timezone and then to Instant. As I understand it, you would get wrong time if the system timezone is not UTC and the time didn't contain a timezone (which would be OK for instant since it is implicitly always in UTC)

**Comment by smootoo** [comment]
I wrote this as you were making more comments. In summary though, I think the timestamp is mostly the wrong datatype for Instants as DST shifts may be applied to it depending on configuration.

The timestamp (without time zone) database type is handled different in different db backends (and jdbc implementations). Postgres for example stores timestamps as UTC and so they are the datatype for Instants.
Depending on timezone configurations in the database backend and on the client JVM, DST may be applied to the timestamp. In a timezone where DST is in effect, then an hour may be gained or lost during the changeover. For example, in ‘London/Europe’, if you try to write the instant ‘2018-03-25T01:37:40Z’ into a timestamp column, it will come back from the db as ‘2018-03-25T02:37:40Z’.
This is acceptable and probably intended for LocalDateTime, but Instants need to be continuous and always increasing, with a larger value being later on the timeline.
Where timestamps with time zone are supported by the db backend, then writing Instants as timestamps with timezone ‘UTC’ or offset ‘+00:00’ gives us the characteristics we need.

I’ll write up something like the above in a big comment so that people know what’s going on and why.

Reviewing this PR is the time to pick up on this though, so thanks for looking. We now have a comprehensive set of tests try alternative implementations to make sure there are no edge cases where values aren’t round-tripped correctly.

**Comment by smootoo** [comment]
I agree with you that for Instants, `2018-03-25T01:37:40Z` must stay unchanged whatever the locally configured time zone is.
But if that same Instant is converted into a `LocalDateTime` (i.e. a type without a timezone) and that is stored in a database that is configured with a local time zone of `London/Europe` into a database column of type `TIMESTAMP` then because this time is in the local DST gap (the clocks have gone forward an hour), the db will turn that time into `2018-03-25T02:37:40`. This is what I meant by probably intended behaviour for `LocalDateTime`, that any locally enforced DST gaps are applied, but the value still doesn't have a timezone.

**Comment by hvesalai** [comment]
Ah ok, so then we need just big block letters saying *Caveat emptor DST*

**Comment by smootoo** [comment]
Indeed. There are dragons here everywhere! Like I said, I'll stick in a big comment explaining what's going on. I personally couldn't think of a use for LocalDateTime that isn't better served with another time type. But I guess it supports some legacy behaviours.

**Comment by hvesalai** [comment]
Instead of a big comment, should we invest in a new page on the documentation?

**Comment by smootoo** [comment]
Sure. I'll draft something at the weekend. OK?

**Comment by hvesalai** [comment]
Yes, that sounds good!

**Comment by hvesalai** [comment]
Also, if you can say when the code is ready for a new round of code-review, I can then review and add comments.

**Comment by smootoo** [comment]
Sorry I updated this later that I had planned. Real life got in the way over the last week!

I added a page of documentation. The code is ready for review now.

The PR has been open in various forms for a long time and has had many different contributors. A fresh pair for eyes would good now. Appreciate any comments you may have.

**Comment by hvesalai** [comment]
@smootoo can you rebase your PR, there is some conflicts

File: slick/src/main/scala/slick/jdbc/H2Profile.scala [hunk 12]
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
       override def valueToSQLLiteral(value: UUID) = "'" + value + "'"
       override def hasLiteralForm = true
     }
+    override val instantType : InstantJdbcType = new InstantJdbcType {
+      val formatter = new DateTimeFormatterBuilder()
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/H2Profile.scala]: Can you add a comment why we need the formatter here. DB2 and Derby use `Instant.parse`, which can't this?

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala [hunk 13]
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
...
       override def sqlTypeName(sym: Option[FieldSymbol]) = "BINARY(16)"
     }
+    override val localTimeType = new LocalTimeJdbcType {
+      override def sqlType = java.sql.Types.TIME
+
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIME(3)"
+
+      override def getValue(r: ResultSet, idx: Int): LocalTime = {
+        // Fixes an issue caused because Hsqldb outputs the time in a non-standard format
+        // not adding trailing zeros to the hours. For example: '2:14:41.421' instead of '02:14:41.421'
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: comment speaks about "trailing" (after something), maybe it means "leading" (before something)

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala [hunk 14]
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
...
+        r.getString(idx) match {
+          case null => null
+          case isoTime if isoTime.indexOf(':') == 2 => LocalTime.parse(isoTime)
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: Why is the formatter on line 162 not used?

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala [hunk 15]
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
...
+          .append(DateTimeFormatter.ISO_LOCAL_DATE)
+          .appendLiteral(' ')
+          .append(timeFormatter)
+          .toFormatter()
+      }
+
+      /**
+        * Add offset padding e.g. +1:00 becomes +01:00.
+        */
+      def offsetConvertHsqldbToISO(hsqldbString : String) : String = hsqldbString.takeRight(5).toArray match {
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: Why are these needed? They look a bit suspicious to me. Couldn't you just use a `DateTimeFormatter` instead? 

Please add a comment on why a `DateTimeFormatter` cannot be used.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala [hunk 16]
```diff
@@ -324,13 +335,154 @@ trait OracleProfile extends JdbcProfile {
...
+          case timestamp => timestamp.toLocalDateTime.toLocalTime
+        }
+      }
+    }
+
+    // LocalDateTime and Instant are the 2 types which have no TZ component
+    // So, store them at UTC timestamps, otherwise the JDBC layer might attempt to map them
+    // and with DST changes, there are some times which will be unrepresentable during the switchover
+    class LocalDateTimeJdbcType extends super.LocalDateTimeJdbcType {
+//      private[this] val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS x")
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/OracleProfile.scala]: remove the commented out code if there is no reason to keep it. If there is a reason, document it.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala [hunk 17]
```diff
@@ -404,3 +556,180 @@ object OracleProfile extends OracleProfile {
     case class AutoIncTriggerName(name: String) extends ColumnOption[Nothing]
   }
 }
+
+
+/* The following code will be used when the native implementation of
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/OracleProfile.scala]: what does this comment mean? "native implementation of ... are natively implemented"

File: slick/src/main/scala/slick/jdbc/SQLServerProfile.scala [hunk 18]
```diff
@@ -248,29 +255,143 @@ trait SQLServerProfile extends JdbcProfile {
...
+        r.getString(idx) match {
+          case null =>
+            null
+          case dateStr =>
+            OffsetDateTime.parse(dateStr, formatter).toInstant()
+        }
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        value match {
+          case null =>
```
Comment (hvesalai) [inline slick/src/main/scala/slick/jdbc/SQLServerProfile.scala]: this is the only profile that handles `value == null`. Why don't the other do it?

**Comment by hvesalai** [comment]
Ok. Now there is a plethora of comments. Can you go through them all and reply / make changes as needed.

Also it would be very good if @tminglei had time to read the Postgres code here.

File: slick/src/main/scala/slick/jdbc/H2Profile.scala (hunk 12 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/H2Profile.scala]: Comment added

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala (hunk 13 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: Comment fixed. You're right. Is should have been leading.

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala (hunk 14 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: The formatter on line 162 is for offset times. It has a non-optional offset component and so can't be used for LocalTime. The default LocalTime parser (which is used here) works fine though.

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala (hunk 15 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: The comment on lines 157 to 159 explain why these are needed in addition to a `DateTimeFormatter` This whole trait is for handling that the offsets don't have leading zeros.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala (hunk 16 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/OracleProfile.scala]: Commented out code removed

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala (hunk 17 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/OracleProfile.scala]: That comment was left over from one of the intermediate commits. Removed.

**Comment by smootoo** [comment]
I can't reply directly to your comment on the DB2Profile.
I don't understand that comment. Could you elaborate?

File: doc/src/datetimetypes.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+The [java.time.* types ](https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html) that were introduced in Java 8, are well thought out and comprehensive. Mapping these into the disparate date and time
+types implemented by the different database backends and mediated through the various jdbc implementations is tricky.
+
+There are two principles that have been applied in mapping the java.time types into database types.
+
+1. The values should be unchanged by a roundtrip to and from the database.
+That means that you get back the same value that was stored, irrespective of the configuration of the
+database or the JVM.
+
+2. It should be clear to you as a user of the database what those values are. This may be with an appropriate database type, if one is available. If not, a varchar with a "stringified" value will be used.
```
[the remark this answers is not in the document]
Comment (smootoo) [reply doc/src/datetimetypes.md]: Removed my made up word.
Elaborated on the comparability of the string values.

File: doc/src/datetimetypes.md [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+
+The different java.time types have important characteristics. E.g `Instant`s need to be continuous and always increasing,
+with a larger value being later on the timeline.
+The timestamp (without time zone) database type is handled different in different db backends (and jdbc implementations). Postgres, for example, stores timestamps as UTC and so they are approproriate to be the datatype for Instants. Depending on timezone configurations in the database backend and on the client JVM, DST may be applied
```
[the remark this answers is not in the document]
Comment (smootoo) [reply doc/src/datetimetypes.md]: I did mean for db others than postgres. Updated.
Added explanation for DST and why it's a problem.

File: slick/src/main/scala/slick/jdbc/SQLServerProfile.scala (hunk 18 above)
Comment (smootoo) [reply slick/src/main/scala/slick/jdbc/SQLServerProfile.scala]: The `case null =>` couldn't be reached. They've been removed.

**Comment by smootoo** [comment]
@hvesalai I've updated based on your comments. Thanks.
I don't think this needs a rebase. I can merge as is, cleanly.

File: slick/src/main/scala/slick/jdbc/HsqldbProfile.scala (hunk 15 above)
Comment (hvesalai) [reply slick/src/main/scala/slick/jdbc/HsqldbProfile.scala]: Ok

**Comment by hvesalai** [comment]
Added two more requests to review coments.

File: slick-testkit/src/codegen/scala/slick/test/codegen/GenerateRoundtripSources.scala [hunk 21]
```diff
@@ -79,7 +79,7 @@ class Tables(val profile: JdbcProfile){
   class X(tag: Tag) extends Table[(Int,Int,Option[Int],Int,Double,String,Option[Int],Option[Int],Option[String],Option[String],Option[String])](tag, "X") {
     def pk = column[Int]("pk")
     def pk2 = column[Int]("pk2")
-    def pkpk = primaryKey( "", (pk,pk2) ) // pk column collision
+    def pkpk = primaryKey( "primary_key2", (pk,pk2) ) // pk column collision
```
[the remark this answers is not in the document]
Comment (smootoo) [reply slick-testkit/src/codegen/scala/slick/test/codegen/GenerateRoundtripSources.scala]: PR #1891 has been raised

File: doc/src/datetimetypes.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+Where timestamps with time zone are supported by the db backend, then writing Instants as timestamps with timezone
+`UTC` or offset `+00:00` gives us the characteristics we need for an `Instant`.
+
+
+One java.time type to be careful of is `LocalDateTime`, that behaves most like the legacy `java.sql.Timestamp` and is mapped into a Timestamp database type. For these, the exact behaviour for DST related can depend on the environment configuration of the database or the JVM. And so is the one type that can break the "roundtripping" rule if the value being stored falls into a DST gap for a given environment configuration.
+
+For example, if your environment is configured to be `Europe/London`, and you try to write the `LocalDateTime` `2018-03-25T01:37:40` into a database that performs DST shifts, such as Oracle, it will come back from the db as `2018-03-25T02:37:40`. This may be the desired behaviour, and is what happens with some databases for a Timestamp without timezone type. If this isn't the desired behaviour, then using a more appropriate type, e.g. `Instant` or `ZonedDateTime` will not have a DST gap issues, irrespective of environment configuration.
+
```
[the remark this answers is not in the document]
Comment (smootoo) [reply doc/src/datetimetypes.md]: Documentation updated

**Comment by hvesalai** [comment]
Almost there. Now if you can squash all the commits to one and rebase, then I can hit the merge button. Now this PR still contains the same changes as in #1891, but after a rebase they should be gone.

**Comment by hvesalai** [comment]
Here goes the planet! Thank you!

**Comment by nafg** [comment]
Well... this broke my app, and not at compile time...

I have different columns of different SQL types in various places that are mapped to Instant. I don't see a workaround other than doing one of the following (which I probably should do anyway, but that doesn't mean I would prefer less overhead right now):

1. Convert the old column types
2. Use something like a tagged type, so that the column type includes the underlying type somehow but values can still be used as the mapped type (essentially along the lines of `Instant with Phantom[Long]`)

**Comment by hvesalai** [comment]
@nafg there was somebody with identical problem I think. Just can't remember where that discussion is.
